client/rest-high-level/src/main/java/org/elasticsearch/client/security/AuthenticateResponse.java

```diff
@@ -42,12 +46,24 @@
     static final ParseField REALM_NAME = new ParseField("name");
     static final ParseField REALM_TYPE = new ParseField("type");
     static final ParseField AUTHENTICATION_TYPE = new ParseField("authentication_type");
+    static final ParseField TOKEN = new ParseField("token");
 
     @SuppressWarnings("unchecked")
     private static final ConstructingObjectParser<AuthenticateResponse, Void> PARSER = new ConstructingObjectParser<>(
             "client_security_authenticate_response", true,
-            a -> new AuthenticateResponse(new User((String) a[0], ((List<String>) a[1]), (Map<String, Object>) a[2],
-                (String) a[3], (String) a[4]), (Boolean) a[5], (RealmInfo) a[6], (RealmInfo) a[7], (String) a[8]));
+            a -> {
+                if (a[9] == null) {
+                    return new AuthenticateResponse(new User((String) a[0], ((List<String>) a[1]), (Map<String, Object>) a[2],
+                        (String) a[3], (String) a[4]), (Boolean) a[5], (RealmInfo) a[6], (RealmInfo) a[7], (String) a[8]);
+                } else {
+                    final AuthenticateResponse response = new AuthenticateResponse(
+                        new User((String) a[0], ((List<String>) a[1]), (Map<String, Object>) a[2],
+                        (String) a[3], (String) a[4]), (Boolean) a[5], (RealmInfo) a[6], (RealmInfo) a[7], (String) a[8],
+                        Map.of(TOKEN_NAME_FIELD, ((Map<String, Object>) a[9]).get(NAME_FIELD)));
```

Comment:
I don't think this is right. The JSON response exposes `token` as a field, so the Java response object should too.

Reply:
It's certainly weird. But the `AuthenticateResponse` class is pretty weird already. It's internal structure (fields) mimics the server side `Authenticate`, i.e. both have:

* User object
* Auth realm
* Lookup realm
* Auth type

But the issue is that server side `Authenticate` has a quite different serialisation (JSON) form than its internal structure. This means, the client side `AuthenticateResponse` needs to bridge this difference by declaring a [parser](https://github.com/elastic/elasticsearch/blob/3a9ce5cded414fe30dd1511aa5850e5a287ed28e/client/rest-high-level/src/main/java/org/elasticsearch/client/security/AuthenticateResponse.java#L52-L64) that has the shape of the JSON form, but a [constructor](https://github.com/elastic/elasticsearch/blob/7.13/client/rest-high-level/src/main/java/org/elasticsearch/client/security/AuthenticateResponse.java#L47-L50) to restore the internal structure. 

I was trying to fit my changes with a similar style. I agree it's awkward and will change it to be a top level `token` field. But we might want to revamp `AuthenticateResponse` so that all other fields are also aligned with the JSON response from the server side, instead of the backing class's internal strcuture.